packages/cardano-services-client/src/WebSocket.ts

```diff
@@ -0,0 +1,182 @@
+
+  private heartbeatInterval: number;
+  private heartbeatTimeout: NodeJS.Timeout;
+  private logger: Logger;
+  private status: WSStatus = 'idle';
+  private url: URL;
+  private ws: WebSocket;
+
+  private networkInfoSubjects = {} as {
+    [m in NetworkInfoMethods]: ReplaySubject<AsyncReturnType<NetworkInfoProvider[m]>>;
+  };
+
+  constructor(deps: WsClientDependencies, cfg: WsClientConfiguration) {
+    this.heartbeatInterval = (cfg.heartbeatInterval || 55) * 1000;
+    this.logger = deps.logger;
+    this.url = cfg.url;
+
+    this.networkInfoSubjects = {
+      eraSummaries: new ReplaySubject<EraSummary[]>(1),
+      genesisParameters: new ReplaySubject<Cardano.CompactGenesis>(1),
+      ledgerTip: new ReplaySubject<Cardano.Tip>(1),
+      lovelaceSupply: new ReplaySubject<SupplySummary>(1),
+      protocolParameters: new ReplaySubject<Cardano.ProtocolParameters>(1),
+      stake: new ReplaySubject<StakeSummary>(1)
+    };
+    this.networkInfo = this.networkInfoSubjects;
+
+    this.networkInfoProvider = {
+      eraSummaries: () => firstValueFrom(this.networkInfo.eraSummaries),
+      genesisParameters: () => firstValueFrom(this.networkInfo.genesisParameters),
+      healthCheck: () => Promise.resolve({ ok: this.status === 'connected' }),
+      ledgerTip: () => firstValueFrom(this.networkInfo.ledgerTip),
+      lovelaceSupply: () => firstValueFrom(this.networkInfo.lovelaceSupply),
+      protocolParameters: () => firstValueFrom(this.networkInfo.protocolParameters),
+      stake: () => firstValueFrom(this.networkInfo.stake)
+    };
+
+    this.connect();
+  }
+
+  private connect() {
+    this.status = 'connecting';
+
+    const ws = new WebSocket(this.url);
+
+    // eslint-disable-next-line sonarjs/cognitive-complexity
+    ws.onmessage = (event) => {
+      try {
+        if (typeof event.data !== 'string') throw new Error('Unexpected data from WebSocket ');
+
+        const message = NJSON.parse<WSMessage>(event.data);
+        const { clientId, networkInfo } = message;
+
+        if (clientId) this.logger.info(`Connected with clientId ${(this.clientId = clientId)}`);
+
+        if (networkInfo) {
+          const { eraSummaries, genesisParameters, ledgerTip, lovelaceSupply, protocolParameters, stake } = networkInfo;
+
+          if (eraSummaries) this.networkInfoSubjects.eraSummaries.next(eraSummaries);
+          if (genesisParameters) this.networkInfoSubjects.genesisParameters.next(genesisParameters);
+          if (lovelaceSupply) this.networkInfoSubjects.lovelaceSupply.next(lovelaceSupply);
+          if (protocolParameters) this.networkInfoSubjects.protocolParameters.next(protocolParameters);
+          if (stake) this.networkInfoSubjects.stake.next(stake);
+          // Emit ledgerTip as last one
+          if (ledgerTip) this.networkInfoSubjects.ledgerTip.next(ledgerTip);
+        }
+      } catch (error) {
+        this.logger.error(error, 'While parsing message', event.data, this.clientId);
```

Comment:
Same as

Reply:
Also this should never happen. I suggest the same approach: let's spend time on it if it actually happens.